Add memory format argument to the `clone` operator

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27270 Add memory format support to `ones_like` operator
* #27262 Add memory format support to `full_like` operator
* #27244 Add memory format support to `empty_like` operator
* #27228 Add memory format support to typecasting shortcuts `byte`,`char`,`double`,`bool`,`half`,`int`,`long`,`short`,`float`,`bfloat16`
* #27223 Add memory format support to `cpu` and `cuda` operators
* #27107 Add memory_format support `to` and `type` operators
* **#27106 Add memory format argument to the `clone` operator**



Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17699357](https://our.internmc.facebook.com/intern/diff/D17699357)

VitalyFedyunin committed:
Add memory format argument to the `to` operator

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -130,6 +130,35 @@ bool TensorImpl::compute_strides_like_channels_last() const {
+    perm[i] = i;
+  }
+  // Sort by strides, leaving 0 and 1 sized dims at the end of the array
+  std::sort(perm.begin(), perm.end(), [&](int a, int b) {
```
I hope `std::sort` is fast on small arrays :)

ezyang reviewed (approved):
okey

dzhulgakov commented on c10/core/TensorImpl.cpp:
```diff
@@ -130,6 +130,35 @@ bool TensorImpl::compute_strides_like_channels_last() const {
+  }
+  // Sort by strides, leaving 0 and 1 sized dims at the end of the array
+  std::sort(perm.begin(), perm.end(), [&](int a, int b) {
+      if (sizes_[a] < 2) {
```
do we actually want to allow strides 0? If someone created a tensor with stride 0 - isn't it overlapping?

VitalyFedyunin replied:
Strides 0 is not allowed, Size 0 is fine. But generally it is already considered is_contiguous by old code and density not going to be checked.

VitalyFedyunin committed:
Update on "Add memory format argument to the `clone` operator"


Adds memory_format option to the `clone` operator.

Introduce new `clone` behavior if used with `input_t.clone(memory_format=torch.preserve_format)`:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17699357](https://our.internmc.facebook.com/intern/diff/D17699357)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "    Add memory format argument to the `to` operator"




Fix Win compilation error

Differential Revision: [D17699357](https://our.internmc.facebook.com/intern/diff/D17699357)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format argument to the `clone` operator"




Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17699357](https://our.internmc.facebook.com/intern/diff/D17699357)

[ghstack-poisoned]
